### PR TITLE
Masking dev

### PR DESCRIPTION
**Pull request by Sh-imaa**
- supports multi-class (supports one as well depends on the config files)
- clipping borders
- notebooks to run experiments 
- changed in config files

### CODE REVIEW

File: glacier_mapping/data/mask.py [hunk 1]
```diff
@@ -58,9 +60,18 @@ def generate_masks(img_paths, shps_paths, output_base="mask", out_dir=None):
         mask = generate_mask(img.meta, shps)
         out_path = pathlib.Path(out_dir, f"{output_base}_{k:02}")
         np.save(str(out_path), mask)
+        # get borders
```
Comment (krisrs1128) [inline glacier_mapping/data/mask.py:63]: I don't understand why we need new code for this. Can't we just clip the shapefiles to the borders in advance, and use that as input into the previous implementation?

I'm just very wary about changes in the code that will make it less generic to other applications.

File: glacier_mapping/train.py [hunk 2]
```diff
@@ -136,13 +145,30 @@ def log_images(writer, frame, batch, epoch, stage="train"):
...
     if epoch == 0:
-        writer.add_image(f"{stage}/x", make_grid(pm(squash(x))), epoch)
-        writer.add_image(f"{stage}/y", make_grid(pm(y)), epoch)
+        writer.add_image(f"{stage}/x", make_grid(pm(squash(x[:, :, :, :3]))), epoch)
+        writer.add_image(f"{stage}/y", make_grid(y.unsqueeze(1)), epoch)
 
-    writer.add_image(f"{stage}/y_hat", make_grid(pm(y_hat)), epoch)
+    writer.add_image(f"{stage}/y_hat", make_grid(y_hat.unsqueeze(1)), epoch)
+
+def update_metrics(main_metrics, batch_metrics):
```
Comment (krisrs1128) [inline glacier_mapping/train.py:162]: What does this code do?

File: scripts/experiment_notebooks/preprocess.ipynb [hunk 3]
```diff
@@ -0,0 +1,294 @@
+{
+ "cells": [
```
Comment (krisrs1128) [inline scripts/experiment_notebooks/preprocess.ipynb:2]: Are these notebooks different than the `preprocess` or `train` scripts in `scripts/`?

File: conf/permutation_exper/Readme.md [hunk 4]
```diff
@@ -0,0 +1 @@
+Configuration files for channel importance expirement
```
Comment (krisrs1128) [inline conf/permutation_exper/Readme.md]: expirement -> experiment

File: glacier_mapping/train.py [hunk 5]
```diff
@@ -136,13 +145,30 @@ def log_images(writer, frame, batch, epoch, stage="train"):
     Return:
         Images Logged onto tensorboard
     """
-    pm = lambda x: x.permute(0, 3, 2, 1)
+    pm = lambda x: x.permute(0, 3, 1, 2)
     squash = lambda x: (x - x.min()) / (x.max() - x.min())
 
     x, y = batch
-    y_hat = torch.sigmoid(frame.infer(x))
+    if frame.multi_class:
```
Comment (krisrs1128) [inline glacier_mapping/train.py]: This is much less readable than the previous version of train.

I think you could modularize it. Alternatively, I personally don't think we need to support both 1 and multichannel outputs (we can just have multichannel). But should check with others to see if they think it's okay.

**Review by krisrs1128 — CHANGES_REQUESTED** [review]

File: glacier_mapping/data/mask.py (hunk 1 above)
Comment (Sh-imaa) [reply glacier_mapping/data/mask.py:63]: Clipping shp files to borders will create more false +ves. We can crop the original images in GEE though or in a separate step other than masking and slicing. 
The code still runs without borders if it's not supplied (it will just act as before). If that causes concerns about being generic we can still have a separate script for that outside of the library.

File: glacier_mapping/train.py (hunk 2 above)
Comment (Sh-imaa) [reply glacier_mapping/train.py:162]: It logs all predicted classes maps. If we have n classes, it logs n maps. It was logging one channel only before

File: scripts/experiment_notebooks/preprocess.ipynb (hunk 3 above)
Comment (Sh-imaa) [reply scripts/experiment_notebooks/preprocess.ipynb:2]: not majorly, it just takes borders into account. We can edit the scripts to do that as well. I thought we want both notebooks and scripts from previous discussions, but we can keep one version only.

File: glacier_mapping/train.py (hunk 5 above)
Comment (Sh-imaa) [reply glacier_mapping/train.py]: it can be changed to something like frame.pred, it might cause a little duplication though. I changed it a little, you can have a look at it or we can support only 2 classes.

File: glacier_mapping/data/mask.py (hunk 1 above)
Comment (krisrs1128) [reply glacier_mapping/data/mask.py:63]: Ah, I remember this discussion now. Didn't we agree to do this as a separate filtering step within the preprocessing script?

In my view, the entire process of removing the images that are outside of the borders is a preprocessing step, and should not referenced in the training code. The training code should operate on more or less abstract input / output pairs, without needing context dependent knowledge.

It's not enough to make sure the code works when not supplied borders. My concern is not about functionality, it's about clarity and maintainability.

File: glacier_mapping/train.py (hunk 2 above)
Comment (krisrs1128) [reply glacier_mapping/train.py:162]: Sorry, I didn't mean this particular line, but block that you inserted starting from around here. Did we have an update_metrics function before? Or is this new / what is its motivation?

File: scripts/experiment_notebooks/preprocess.ipynb (hunk 3 above)
Comment (krisrs1128) [reply scripts/experiment_notebooks/preprocess.ipynb:2]: It just seems confusing to have both, unless sufficient context is given. Can you make it clear that this is for the permutation experiments, either by putting it in an appropriate folder or creating a readme?

File: glacier_mapping/models/frame.py [hunk 6]
```diff
@@ -144,7 +186,7 @@ def metrics(self, y_hat, y, metrics_opts):
             if "threshold" in metric.keys():
                 y_hat = y_hat > metric["threshold"]
 
-                metric_fun = globals()[k]
-                metric_value = metric_fun(y_hat, y)
-            results[k] = np.mean(np.asarray(metric_value))
+            metric_fun = globals()[k]
+            metric_value = metric_fun(y_hat, y)
```
Comment (krisrs1128) [inline glacier_mapping/models/frame.py]: is `metric_value` used again later? if not, just use the output of the function directly in `results[k]`

File: glacier_mapping/models/frame.py [hunk 7]
```diff
@@ -98,6 +105,35 @@ def infer(self, x):
         with torch.no_grad():
             return self.model(x).permute(0, 2, 3, 1)
 
+    def predict(self, y_hat):
```
Comment (krisrs1128) [inline glacier_mapping/models/frame.py]: Modularizing like this is much better. Though, can you change the name from `predict` to a more appropriate verb? Predict sounds like you give x as input and it returns y_hat. This is taking a probability map and outputting a segmentation. Maybe something like `self.threshold` or `self.segment` is more appropriate?

File: glacier_mapping/train.py [hunk 8]
```diff
@@ -70,15 +73,18 @@ def validate(loader, frame, metrics_opts):
     :return (val_loss, metrics): A tuple containing the average validation loss
       and the metrics on the validation set.
     """
-    loss, metrics = 0, []
+    loss, metrics = 0, {}
     for x, y in loader:
+        channel_first = lambda x: x.permute(0, 3, 1, 2)
```
Comment (krisrs1128) [inline glacier_mapping/train.py]: This can go outside the for loop. It's a little confusing to have an x defined in the loop as well as in this lambda.

File: glacier_mapping/data/mask.py (hunk 1 above)
Comment (Sh-imaa) [reply glacier_mapping/data/mask.py:63]: I think it being in generate_masks is part of the preprocessing step rather than the training code. But we can leave it out completely and have a code the crop the entire masks without using any preprocessing functionality from the library. it will be a script to run (it might take long because of the extra read). If that's easier to maintain, then it's easy fix.

File: glacier_mapping/train.py (hunk 2 above)
Comment (Sh-imaa) [reply glacier_mapping/train.py:162]: We didn't. We updated metrics directly before by making a dataframe out of a list of dict and take the mean, as we had one value per metric. Since we have multiple classes now, update takes care of updating an array.

File: scripts/experiment_notebooks/preprocess.ipynb (hunk 3 above)
Comment (Sh-imaa) [reply scripts/experiment_notebooks/preprocess.ipynb:2]: It's not for the permutation specifically. It's just adds the target_locs saving (for fixed runs) and add the border bit. I'll remove it if it causes a confusion. 
the current script might need minor changes to reflect the code though.

File: glacier_mapping/data/mask.py (hunk 1 above)
Comment (krisrs1128) [reply glacier_mapping/data/mask.py:63]: Ah you're completely right, I was just having trouble jumping across the contexts in this merge. For some reason I thought this was happening in train, but I'm happy with it happening in `generate_masks`. This part looks good to me, then.

File: glacier_mapping/train.py (hunk 2 above)
Comment (krisrs1128) [reply glacier_mapping/train.py:162]: Hm, okay. It makes sense to create a new function, and it's nice how concise it is. I wish it were a little more obvious what it did, but I can't think of a better way to do it right now.

File: glacier_mapping/train.py [hunk 9]
```diff
@@ -136,13 +145,30 @@ def log_images(writer, frame, batch, epoch, stage="train"):
...
-        writer.add_image(f"{stage}/x", make_grid(pm(squash(x))), epoch)
-        writer.add_image(f"{stage}/y", make_grid(pm(y)), epoch)
+        writer.add_image(f"{stage}/x", make_grid(pm(squash(x[:, :, :, :3]))), epoch)
+        writer.add_image(f"{stage}/y", make_grid(y.unsqueeze(1)), epoch)
 
-    writer.add_image(f"{stage}/y_hat", make_grid(pm(y_hat)), epoch)
+    writer.add_image(f"{stage}/y_hat", make_grid(y_hat.unsqueeze(1)), epoch)
+
+def update_metrics(main_metrics, batch_metrics):
+    for k, v in batch_metrics.items():
```
Comment (krisrs1128) [inline glacier_mapping/train.py:168]: You should probably just loop over values, right? The `k` here is immediately overwritten in the next line.

File: scripts/experiment_notebooks/preprocess.ipynb (hunk 3 above)
Comment (krisrs1128) [reply scripts/experiment_notebooks/preprocess.ipynb:2]: Hm, I don't think it's worth postponing the merge just because of this issue. We should create a future task of reconciling the scripts. I suspect that this notebook supersedes the other scripts, but it would still take some attention to check.

File: glacier_mapping/train.py (hunk 9 above)
Comment (Sh-imaa) [reply glacier_mapping/train.py:168]: not sure I understand. I take k (key) to update the base matrix. It doesn't get overwritten. It is concatenated at the location of k.

Comment (krisrs1128) [reply glacier_mapping/train.py:168]: Ah of course, I'm just going crazy, the next line is an `if` not a `for`.

File: glacier_mapping/train.py (hunk 2 above)
Comment (Sh-imaa) [reply glacier_mapping/train.py:162]: I commented it to make it clearer

**Review by krisrs1128 — APPROVED** [review]
This looks good to me now. Feel free to merge whenever you are ready.
